sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -190,7 +191,36 @@ case class AdaptiveSparkPlanExec(
           executionId.foreach(onUpdatePlan(_, result.newStages.map(_.plan)))
 
           // Start materialization of all new stages and fail fast if any stages failed eagerly
-          result.newStages.foreach { stage =>
+
+          // SPARK-33933: we should materialize broadcast stages first and wait the
+          // materialization finish before materialize other stages, to avoid waiting
+          // for broadcast tasks to be scheduled and leading to broadcast timeout.
+          val broadcastMaterializationFutures = result.newStages
+            .filter(_.isInstanceOf[BroadcastQueryStageExec])
+            .map { stage =>
+            var future: Future[Any] = null
+            try {
+              future = stage.materialize()
+              future.onComplete { res =>
+                if (res.isSuccess) {
+                  events.offer(StageSuccess(stage, res.get))
+                } else {
+                  events.offer(StageFailure(stage, res.failed.get))
+                }
+              }(AdaptiveSparkPlanExec.executionContext)
+            } catch {
+              case e: Throwable =>
+                cleanUpAndThrowException(Seq(e), Some(stage.id))
+            }
+            future
+          }
+
+          // Wait for the materialization of all broadcast stages finish
+          broadcastMaterializationFutures.foreach(ThreadUtils.awaitReady(_, Duration.Inf))
```

Comment:
Is it necessary to wait until all 'BroadcastQueryStageExec' are materialized

Reply:
In deed, there will be a little waste of resources. This is the same behavior as non-AQE. Given the lightweight of broadcast, it should not cause too much time, few seconds in normal. I think that's acceptable.
If not wait, there's still probability that situations in #30998 will occur and cause broadcast timeout.

Reply:
@zhongyu09 It might be better to give a benchmark to compare the performance difference between before and after

Reply:
Yes, do we have some benchmark testing framework?

Reply:
Micro benchmark can base on `BenchmarkBase` or `SqlBasedBenchmark`, like `DataSourceReadBenchmark`. But for this scenario, I prefer to you can give a description of the test process and a comparison of the benchmark numbers, maybe need some screenshot

Reply:
I mean test tools and env like Kubernetes integration test or else, otherwise I need to run the test in my local computer? It seems at least we needs to run TPCDSQueryBenchmark.

Reply:
I agree the perf regression may be small, but the gain is also small. If we simply submit broadcast stages first, in most cases we are already fine. I'm not sure if it worths to introduce perf regression for such a small gain.

Reply:
Agree. But I am ok to do perf regression. I am just learning how to do it.

Reply:
I think people were OK with the original solution, it's just a problem of writing tests. We can add a way to record the stage submission order, then use it to test.

Reply:
But the original solution cannot guarantee the stage submission order. The UT may be flaky. I have tried to retry the test  when fail, and it's always fail after retry.

Reply:
The UT is flaky because we were checking job submission order, while we only guarantee stage submission order. We can update the UT to check stage submission order.

Reply:
You mean the order of calling materialize for QueryStage? It is not recorded yet. As I said, the original solution cannot guarantee the normal stage(not the QueryStage) submission order. And cannot avoid broadcast time out unless we catch the exception. But I don't think that's a good UT.

Reply:
It's great to have a better fix to guarantee the job submission order, but the current one is not (can cause perf regression if the resource is sufficient). Guarantee query stage submission is not great but it's better than doing nothing. If we agree with the compromise, then we should write UT to verify the behavior, which is the query stage submission order.

If you don't like this compromise, let's slow down and think about how to fix the problem without perf regression.

Reply:
I am fine with the partial fix like #30998. I wonder is it too heavy to add new event just for UT? 
I tend to fix the problem without perf regression. But we can also let the partial fix goes first.

Reply:
We can also log the stage submission and then write test to verify the log.

Reply:
That's an idea. I will have a look for how to do this. Do we have any UT to verify the log?

Reply:
yea a lot, e.g. `AdaptiveQueryExecSuite.test log level`

Reply:
Put a partial fix as discussed in https://github.com/apache/spark/pull/31269  cc @viirya